File: python/day07.py
from aoc import *


def conc(a, b): return int(str(a) + str(b))

def is_valid_eq(res, nums, can_use_concat):
    match nums:
        case []: return False
        case [a]: return (a == res) and res
        case [a, *_] if a > res: return False
        case [a, b, *tail]:
            return (is_valid_eq(res, [a+b, *tail], can_use_concat) or
                    is_valid_eq(res, [a*b, *tail], can_use_concat) or
                    (can_use_concat and
                     is_valid_eq(res, [conc(a, b), *tail], can_use_concat)))

def sum_valid(data, can_use_concat):
    return sum(is_valid_eq(res, nums, can_use_concat)
               for res, *nums in data)

def solve(filename):
    data = parse_lines(read_input(filename), ints)
    return [sum_valid(data, concat) for concat in (0, 1)]


solve(7)

File: elixir/day07.exs
defmodule Day07 do
  defp conc(a, b), do: String.to_integer(to_string(a) <> to_string(b))

  defp valid_equation?(res, nums, concat_op?) do
    case nums do
      [] ->
        false

      [a] ->
        a == res

      [a | _] when a > res ->
        false

      [a, b | tail] ->
        ([a + b, a * b] ++ if(concat_op?, do: [conc(a, b)], else: []))
        |> Enum.any?(fn op -> valid_equation?(res, [op | tail], concat_op?) end)
    end
  end

  defp valid?(lines, concat_op?) do
    import Enum

    lines
    |> filter(fn [res | nums] -> valid_equation?(res, nums, concat_op?) end)
    |> map(&hd/1)
    |> sum
  end

  def solve(filename) do
    import AOC

    filename
    |> read_input
    |> parse_lines(:ints)
    |> then(fn x -> {valid?(x, false), valid?(x, true)} end)
  end
end

Day07.solve(7)
